sparse/_coo/numba_extension.py

```diff
@@ -0,0 +1,269 @@
+    def key(self):
+        return self.data_dtype, self.coords_dtype, self.ndim
+
+    @property
+    def data_type(self):
+        return numba.from_dtype(self.data_dtype)[:]
+
+    @property
+    def coords_type(self):
+        return numba.from_dtype(self.coords_dtype)[:, :]
+
+    @property
+    def shape_type(self):
+        return types.UniTuple(types.int64, self.ndim)
+
+    @property
+    def fill_value_type(self):
+        return numba.from_dtype(self.data_dtype)
+
+
+@typeof_impl.register(COO)
+def _typeof_COO(val: COO, c) -> COOType:
+    return COOType(
+        data_dtype=val.data.dtype, coords_dtype=val.coords.dtype, ndim=val.ndim
+    )
+
+
+@register_model(COOType)
+class COOModel(models.StructModel):
+    def __init__(self, dmm, fe_type):
+        members = [
+            ("data", fe_type.data_type),
+            ("coords", fe_type.coords_type),
+            ("shape", fe_type.shape_type),
+            ("fill_value", fe_type.fill_value_type),
+        ]
+        models.StructModel.__init__(self, dmm, fe_type, members)
+
+
+@type_callable(COO)
+def type_COO(context):
+    # TODO: accept a fill_value kwarg
+    def typer(coords, data, shape):
+        return COOType(
+            coords_dtype=numba.numpy_support.as_dtype(coords.dtype),
+            data_dtype=numba.numpy_support.as_dtype(data.dtype),
+            ndim=len(shape),
+        )
+
+    return typer
+
+
+@lower_builtin(COO, types.Any, types.Any, types.Any)
+def impl_COO(context, builder, sig, args):
+    typ = sig.return_type
+    coords, data, shape = args
+    coo = cgutils.create_struct_proxy(typ)(context, builder)
+    coo.coords = coords
+    coo.data = data
+    coo.shape = shape
+    coo.fill_value = context.get_constant_generic(
+        builder, typ.fill_value_type, _zero_of_dtype(typ.data_dtype)
+    )
+    return impl_ret_borrowed(context, builder, sig.return_type, coo._getvalue())
+
+
+@lower_constant(COOType)
+def lower_constant_COO(context, builder, typ, pyval):
+    coords = context.get_constant_generic(builder, typ.coords_type, pyval.coords)
+    data = context.get_constant_generic(builder, typ.data_type, pyval.data)
+    shape = context.get_constant_generic(builder, typ.shape_type, pyval.shape)
+    fill_value = context.get_constant_generic(
+        builder, typ.fill_value_type, pyval.fill_value
+    )
+    return impl_ret_borrowed(
+        context,
+        builder,
+        typ,
+        cgutils.pack_struct(builder, (data, coords, shape, fill_value)),
+    )
+
+
+@contextmanager
+def early_exit_if(builder, stack: ExitStack, cond):
+    """
+    Emit code similar to::
+
+        if (cond) {
+            <body>
+            return;
+        }
+        <everything after this call>
+
+    However, this "return" will just break out of the current `ExitStack`,
+    rather than out of the whole function
+    """
+    then, otherwise = stack.enter_context(builder.if_else(cond, likely=False))
+    with then:
+        yield
+    stack.enter_context(otherwise)
```

Comment:
```python
<iterate through list>
c.builder.store(fail_obj, ret_ptr)
```

Could be here.

Reply:
I'm wary of building too much abstraction here, because @stuartarchibald has suggested that something like this should end up in numba itself. The more specific we get, the harder it will be to gut out this code and replace it with numba stuff if they pick a different level of specificity.

Reply:
I was trying to build more-or-less the equivalent of what happens when a Python statement is executed in the interpretor, but I'm happy either way.

Reply:
Which should be generic enough to end up in Numba.

Reply:
This ends up adding 3 more arguments to this function though (the list, the fail object, the return pointer), which i think obscures clarity.